### ISSUE TITLE
High-level API for multimodality

### ISSUE THREAD

**Issue by remixer-dec** [abetlen/llama-cpp-python#928]
**Is your feature request related to a problem? Please describe.**

Current high-level implementation of multimodality is relying on a specific prompt format.

**Describe the solution you'd like**

Models like [Obsidian](https://huggingface.co/nisten/obsidian-3b-multimodal-q6-gguf/tree/main) work with llama.cpp server and have a different format. It would be nice to have a high-level API for multimodality in llama-cpp-python to be able to pass `image`/`images` as an argument after initializing `Llama()` with all the paths to required extra-models, without relying on a pre-defined prompt format such as `Llava15ChatHandler`.

**Describe alternatives you've considered**
Alternatively, a custom prompt format class that supports images can be implemented, where prompt string is passed as an argument.

**Comment by abetlen** [comment]
I'll consider adding the multi-modal args to the `Llama` class but I'm worried about growing the API surface too much too quickly.

For now, I can definitely implement an Obsidian chat handler and maybe abstract out some of the Llava stuff to make it easier to make this generic in the future.

Also thank you for bringing that model to my attention, definitely have to give it a try now!

**Comment by JoshuaFurman** [comment]
This seems like an apt place to ask this; How can I supply a local image file to use with multimodal images? Looks like the example assumes an image is being hosted at a url somewhere?

**Comment by abetlen** [comment]
@JoshuaFurman good question, I'll update the docs to include this. It works the same as the OpenAI `gpt-4-vision-preview` you can pass the image as a base64 encoded data url:

```python
import base64

def image_to_base64_data_uri(file_path):
    with open(file_path, "rb") as img_file:
        base64_data = base64.b64encode(img_file.read()).decode('utf-8')
        return f"data:image/png;base64,{base64_data}"

# Replace 'file_path.png' with the actual path to your PNG file
file_path = 'file_path.png'
data_uri = image_to_base64_data_uri(file_path)
```

Then just pass that in place of the data url

**Comment by JoshuaFurman** [comment]
Oh fantastic thank you! And this can be done without running the server right? I'm looking to add this directly into an application

**Comment by abetlen** [comment]
Yup, you just need  to pass the `Llava15ChatHandler` to `Llama` like in the README example so that the class knows how to format the chat with images.

**Comment by JoshuaFurman** [comment]
Great thanks so much! Happy thanksgiving :)

**Comment by XcerontangX** [comment]
> @JoshuaFurman good question, I'll update the docs to include this. It works the same as the OpenAI `gpt-4-vision-preview` you can pass the image as a base64 encoded data url:
> 
> ```python
> import base64
> 
> def image_to_base64_data_uri(file_path):
>     with open(file_path, "rb") as img_file:
>         base64_data = base64.b64encode(img_file.read()).decode('utf-8')
>         return f"data:image/png;base64,{base64_data}"
> 
> # Replace 'file_path.png' with the actual path to your PNG file
> file_path = 'file_path.png'
> data_uri = image_to_base64_data_uri(file_path)
> ```
> 
> Then just pass that in place of the http url

Thanks this works

**Comment by abetlen** [comment]
Closed this by mistake, though will be solved in #1147
